### ISSUE TITLE
False positive: test is passing when running specific test file, but failing when running directory

### ISSUE THREAD

**Issue by dbrekelmans** [sebastianbergmann/phpunit#4456]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 9.3.8
| PHP version         | 7.4.9
| Installation Method | Composer

#### Summary

Test is passing when running specific test file, but failing when running directory.

#### Current behavior

Running `vendor/bin/phpunit tests/Browser/VersionResolverFactoryTest.php` is passing, but `vendor/bin/phpunit tests/` is failing with the following message:
```
There were 2 errors:

1) DBrekelmans\BrowserDriverInstaller\Tests\Browser\VersionResolverFactoryTest::testSupportedVersionResolverIsReturned
TypeError: Argument 1 passed to DBrekelmans\BrowserDriverInstaller\Browser\VersionResolverFactory::register() must implement interface DBrekelmans\BrowserDriverInstaller\Browser\VersionResolver, instance of A given, called in /home/runner/work/browser-driver-installer/browser-driver-installer/tests/Browser/VersionResolverFactoryTest.php on line 50

/home/runner/work/browser-driver-installer/browser-driver-installer/src/Browser/VersionResolverFactory.php:36
/home/runner/work/browser-driver-installer/browser-driver-installer/tests/Browser/VersionResolverFactoryTest.php:50

2) DBrekelmans\BrowserDriverInstaller\Tests\Browser\VersionResolverFactoryTest::testFirstSupportedVersionResolverIsReturned
TypeError: Argument 1 passed to DBrekelmans\BrowserDriverInstaller\Browser\VersionResolverFactory::register() must implement interface DBrekelmans\BrowserDriverInstaller\Browser\VersionResolver, instance of A given, called in /home/runner/work/browser-driver-installer/browser-driver-installer/tests/Browser/VersionResolverFactoryTest.php on line 68

/home/runner/work/browser-driver-installer/browser-driver-installer/src/Browser/VersionResolverFactory.php:36
/home/runner/work/browser-driver-installer/browser-driver-installer/tests/Browser/VersionResolverFactoryTest.php:68
```

#### How to reproduce

- Check out https://github.com/dbrekelmans/browser-driver-installer/pull/14
- Run `composer install`
- Run  `vendor/bin/phpunit tests/Browser/VersionResolverFactoryTest.php`
- Run `vendor/bin/phpunit tests`

#### Expected behavior

Both `vendor/bin/phpunit tests/Browser/VersionResolverFactoryTest.php` and `vendor/bin/phpunit tests` should pass

#### Additional information

I set a custom class name on the mock and add it to my factory
```
$versionResolverA = $this->getMockBuilder(VersionResolver::class)->setMockClassName('A')->getMock();
$factory->register($versionResolverA);
```

The register method that gives the type error looks as follows:
```
public function register(VersionResolver $versionResolver) : void
{
    $this->versionResolvers[get_class($versionResolver)] = $versionResolver;
}
```

The type error is a false postive, probably due to the underlying code using the custom classname `A`.
`register() must implement interface VersionResolver, instance of A given`

**Comment by dbrekelmans** [comment]
I have found the cause of this:
`->setMockClassName('A')` is called in both `PathResolverFactoryTest` and `VersionResolverFactoryTest`.
If both tests are run, it appears that the mock from the first test class is returned in the second test class.

Are these custom mock classnames parhaps stored globally @sebastianbergmann?

The issue is resolved if I make the custom class names unique.

**Comment by sebastianbergmann** [comment]
> Are these custom mock classnames perhaps stored globally @sebastianbergmann?

No matter which API you use – `createStub()`, `createMock()`, ..., or `getMockBuilder()->`getMock()` –, PHPUnit has to generate code for a class that extends the class or implements the interface you want to double. By default, PHPUnit uses a random class name for the generated class. You are using `setMockClassName()` to specify the name of the generated class. PHPUnit assumes that you know what you are doing when you use `setMockClassName()`, especially if you use the same name in different tests.

TL;DR: This is working as expected. My advice is to not use custom class names when you do not need them.
